cristin-project/src/main/java/no/unit/nva/cristin/projects/common/CristinQueryBuilder.java

```diff
@@ -0,0 +1,613 @@
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import no.unit.nva.cristin.common.Utils;
+import no.unit.nva.cristin.projects.model.nva.ProjectStatus;
+import nva.commons.apigateway.RequestInfo;
+import nva.commons.apigateway.exceptions.BadRequestException;
+
+/**
+ * Builds Cristin query parameters using builder methods and NVA input parameters.
+ */
+public class CristinQueryBuilder {
+
+    private static final String PARAMETER_PAGE_DEFAULT_VALUE = "1";
+    private static final String PARAMETER_PER_PAGE_DEFAULT_VALUE = "5";
+    private final transient Set<String> invalidKeys = new HashSet<>(0);
+    private final transient CristinQuery cristinQuery;
+
+    /**
+     * Constructor of CristinQuery.Builder.
+     * <p>Usage:</p>
+     * <samp>new CristinQuery.Builder()<br>
+     * .fromRequestInfo(requestInfo)<br>
+     * .withRequiredParameters(IDENTITY,PAGE_CURRENT,PAGE_ITEMS_PER_PAGE)<br>
+     * .validate()<br>.build()
+     * </samp>
+     */
+    public CristinQueryBuilder() {
+        cristinQuery = new CristinQuery();
+    }
+
+    /**
+     * Builder of CristinQuery.
+     */
+    public CristinQuery build() {
+        return cristinQuery;
+    }
+
+    /**
+     * Validator of CristinQuery.Builder.
+     * @throws BadRequestException if parameters are invalid or missing
+     */
+    public CristinQueryBuilder validate() throws BadRequestException {
+        assignDefaultValues();
+        for (var entry : cristinQuery.pathParameters.entrySet()) {
+            throwInvalidPathValue(entry.getKey(), entry.getValue());
+        }
+        for (var entry : cristinQuery.queryParameters.entrySet()) {
+            throwInvalidParamererValue(entry.getKey(), entry.getValue());
+        }
+        if (!requiredMissing().isEmpty()) {
+            throw new BadRequestException(requiredMissingMessage(getMissingKeys()));
+        }
+        if (oneOrMoreOptionalIsMissing()) {
+            throw new BadRequestException(validQueryParameterNamesMessage(validKeys()));
+        }
+        if (!invalidKeys.isEmpty()) {
+            throw new BadRequestException(validQueryParameterNamesMessage(validKeys()));
+        }
+        return this;
+    }
+
+
+    /**
+     * Adds query and path parameters from requestInfo.
+     */
+    public CristinQueryBuilder fromRequestInfo(RequestInfo requestInfo) {
+        return fromPathParameters(requestInfo.getPathParameters())
+                   .fromQueryParameters(requestInfo.getQueryParameters());
+    }
+
+    /**
+     * Adds parameters from path.
+     * */
+    public CristinQueryBuilder fromPathParameters(Map<String, String> parameters) {
+        parameters.forEach(this::setPath);
+        return this;
+    }
+
+    /**
+     * Adds parameters from query.
+     */
+    public CristinQueryBuilder fromQueryParameters(Map<String, String> parameters) {
+        parameters.forEach(this::setValue);
```

Comment:
So this does the magic of only setting query parameters which exists in the request right?

Reply:
correct.

Reply:
`
    private void setValue(String key, String value) {
        var qpKey = keyFromString(key,value);
        switch (qpKey) {
`
This code identify param names and sets value if key is valid
if param name is invalid it is added to invalid keys which is checked in `validate()`

Reply:
```
        var cristinQuery =
            CristinQuery.builder()
                .fromRequestInfo(requestInfo)
                .withRequiredParameters(PATH_ORGANISATION, PATH_PROJECT, LANGUAGE, PAGE_CURRENT, PAGE_ITEMS_PER_PAGE)
                .asNvaQuery()
                .validate()
                .build();
```

- create builder
- add params and path variables
- set required params
- set default to NVA output (maybe it should be default)
- validates (if invalidated exception is thrown (in this step only))
- build should always succeed
result CristinQuery is always defined and correct.

Reply:
```
public CristinQueryBuilder validate() throws BadRequestException {
    assignDefaultValues();
    for (var entry : cristinQuery.pathParameters.entrySet()) {
        throwInvalidPathValue(entry.getKey(), entry.getValue());
    }
    for (var entry : cristinQuery.queryParameters.entrySet()) {
        throwInvalidParamererValue(entry.getKey(), entry.getValue());
    }
    if (!requiredMissing().isEmpty()) {
        throw new BadRequestException(requiredMissingMessage(getMissingKeys()));
    }
    if (oneOrMoreOptionalIsMissing()) {
        throw new BadRequestException(validQueryParameterNamesMessage(validKeys()));
    }
    if (!invalidKeys.isEmpty()) {
        throw new BadRequestException(validQueryParameterNamesMessage(validKeys()));
    }
    return this;
}
```

Reply:
`oneOrMoreOptionalIsMissing `& `invalidKeys `returns the same exception, but have separated checks for clarity